tests/entrypoints/test_controller.py

```diff
@@ -0,0 +1,367 @@
+From controller's perspective, its main duty is to maintain unique central
+entrance place, a.k.a service URL, and dispatch request to workers.
+
+To accomplish this job, it need to:
+    1. make sure controller could correctly receive worker's join request
+    2. make sure worker would repeatedly sending heart beat signal to controller
+    3. make sure remove stale worker if not receive worker's heart beat
+
+And by worker registration, it also could do the job as:
+    1. autoscale with new worker of the same serving model join in.
+    2. model roll update by gradudtely join new model pod while delete old one.
+    3. request load balance by inspect worker's engine pending request len.
+"""
+import json
+import os
+import subprocess
+import sys
+import time
+
+import openai  # use the official client for correctness check
+import pytest
+# using Ray for overall ease of process management, parallel requests,
+# and debugging.
+import ray
+import requests
+from huggingface_hub import snapshot_download
+
+from vllm.utils import get_ip
+
+MAX_SERVER_START_WAIT_S = 600  # wait for server to start for 60 seconds
+# any model with a chat template should work here
+MODEL_NAME = "HuggingFaceH4/zephyr-7b-beta"
+# technically this needs Mistral-7B-v0.1 as base, but we're not testing
+# generation quality here
+LORA_NAME = "typeof/zephyr-7b-beta-lora"
+
+# make check fast
+CONTROLLER_HEART_BEAT_EXPIRATION = 10
+os.environ["VLLM_CONTROLLER_HEART_BEAT_EXPIRATION"] = str(
+    CONTROLLER_HEART_BEAT_EXPIRATION)
+os.environ["VLLM_WORKER_HEART_BEAT_INTERVAL"] = "5"
+host_ip = get_ip()
+controller_addr = "{}:8000".format(host_ip)
+worker1_addr = "{}:8001".format(host_ip)
+worker2_addr = "{}:8002".format(host_ip)
+
+pytestmark = pytest.mark.asyncio
+
+
+@ray.remote
+class Controller:
+
+    def __init__(self, args):
+        env = os.environ.copy()
+        env["PYTHONUNBUFFERED"] = "1"
+        self.proc = subprocess.Popen(
+            ["python3", "-m", "vllm.entrypoints.openai.controller"] + args,
+            env=env,
+            stdout=sys.stdout,
+            stderr=sys.stderr,
+        )
+        self._wait_for_server()
+
+    def ready(self):
+        return True
+
+    def _wait_for_server(self):
+        # run health check
+        start = time.time()
+        while True:
+            try:
+                if requests.get("http://{}/health".format(
+                        controller_addr)).status_code == 200:
+                    break
+            except Exception as err:
+                if self.proc.poll() is not None:
+                    raise RuntimeError("Server exited unexpectedly.") from err
+
+                time.sleep(0.5)
+                if time.time() - start > MAX_SERVER_START_WAIT_S:
+                    raise RuntimeError(
+                        "Server failed to start in time.") from err
+
+    def __del__(self):
+        if hasattr(self, "proc"):
+            self.proc.terminate()
+
+
+@ray.remote(num_gpus=1)
+class Worker:
```

Comment:
To avoid code duplication, consider modifying the existing `tests.utils.Worker` and importing that instead.

Reply:
Done